kong/pdk/cluster.lua

```diff
@@ -0,0 +1,85 @@
+--- Node-level utilities
+--
+-- @module kong.node
+
+local utils = require "kong.tools.utils"
+
+
+local kong = kong
+local CLUSTER_ID_CACHE_KEY = require("kong.constants").CLUSTER_ID_CACHE_KEY
+
+
+local function fetch_cluster_id()
+  local res, err = kong.db.parameters:select({ key = CLUSTER_ID_CACHE_KEY, })
+  if err then
+    return nil, err
+  end
+
+  if res then
+    return res.value
+  end
+
+  local cluster_id = utils.uuid()
+
+  res, err = kong.db:cluster_mutex(CLUSTER_ID_CACHE_KEY, nil, function()
+    -- another worker might have already generated the cluster ID, double check
+    -- in here before overwriting it
+
+    res, err = kong.db.parameters:select({ key = CLUSTER_ID_CACHE_KEY, })
+    if err then
+      error(err)
+    end
+
+    if not res then
+      assert(kong.db.parameters:insert({ key = CLUSTER_ID_CACHE_KEY,
+                                         value = cluster_id, }))
+
+    else
+      cluster_id = res.value
+    end
+  end)
+  if err then
+    error(err)
+  end
+
+  return cluster_id
+end
+
+
+local function new(self)
+  local _CLUSTER = {}
+
+
+  ---
+  -- Returns the unique id for this Hybrid mode cluster. If Kong
+  -- is not running in Hybrid mode, then this method returns nil.
+  --
+  -- All Control Planes and Data Planes belonging to the same
+  -- cluster returns the same cluster ID.
+  --
+  -- @function kong.cluster.get_id
+  -- @treturn string The v4 UUID used by this cluster as its id
+  -- @usage
+  -- local id = kong.cluster.get_id()
+  function _CLUSTER.get_id()
+    if kong.configuration.role == "data_plane" then
+      -- in DB-less mode, parameters are exported and sent to DP
+      local res, err = kong.db.parameters:select({ key = CLUSTER_ID_CACHE_KEY, })
+      if res then
+        return res.value
+      end
+
+      return nil, err
+    end
+
+    return kong.core_cache:get(CLUSTER_ID_CACHE_KEY, nil, fetch_cluster_id)
+  end
```

Comment:
What happens if kong is running in the standalone DB-less mode?

Reply:
Yes I guess in this case we need to add a guard to return `nil` if user did not pre-define a cluster ID in the declarative config file. I will update this PR to check for that. Good catch!

Reply:
@dndx wouldn't it be better to then autogenerate the cluster id? I mean that with dbless the single node is the cluster.

Reply:
@bungle I felt it is not necessary, since in that case we already have `node_id` which is autogenerated, generating another random cluster ID will make data analysis more involved later.

Reply:
> wouldn't it be better to then autogenerate the cluster id? I mean that with dbless the single node is the cluster.

In db less the user can set a cluster id in the config, because only the user knows if there are more nodes running the same config. Or alternatively we should assume there is my cluster, and hence return nil here. So I feel for @dndx s suggestion

Reply:
@Tieske, I am fine with it, I proposed the auto-generation would just be fallback if not specified in config.  Why would `nil` be better than `unique id`. I cannot easily think upsides. Only downsides. E.g. nil is magic value, that we are now attaching meaning. It is also error prone in code if the code does not explicitly check for `nil`ness.

Reply:
Offline we discussed also having `00000000-0000-0000-0000-000000000000` as the default (aka the `nil`), but it was decided to just use `nil`.

Reply:
After discussing we will keep with the current design of returning `nil` when node is not running in a cluster. We can always revisit this later if needed.